gobblin-modules/gobblin-kafka-common/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -591,11 +602,33 @@ private void startAtLatestOffset() {
     @Override
     public void run() {
       try (Timer.Context context = metricContext.timer(WORK_UNITS_FOR_TOPIC_TIMER).time()) {
+        // use shared client if configure, otherwise create a thread local one
+        if (KafkaSource.this.sharedKafkaConsumerClient != null) {
+          KafkaSource.this.kafkaConsumerClient.set(KafkaSource.this.sharedKafkaConsumerClient);
+        } else {
+          KafkaSource.this.kafkaConsumerClient.set(
```

Comment:
Yes, it is one per topic, but set as a thread local to remove contention. The workunit creation time is more than 10 times faster than without this fix, so the instantiation overhead seems minimal when compared to the reduced contention.
Created workunits for 1706 topics in 20 seconds.

I'll create a pool of clients and set and return them within the run() method to address this concern, but it looks like the overhead is fairly minimal.